File: backends/spring-java/src/main/java/dev/stackverse/backend/bookmark/BookmarkCursor.java
package dev.stackverse.backend.bookmark;

import dev.stackverse.backend.common.BadRequestProblem;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Base64;
import java.util.UUID;

public record BookmarkCursor(Instant createdAt, UUID id) {
    public String encode() {
        return Base64.getUrlEncoder()
            .withoutPadding()
            .encodeToString((createdAt + "|" + id).getBytes(StandardCharsets.UTF_8));
    }

    public static BookmarkCursor of(Bookmark bookmark) {
        return new BookmarkCursor(bookmark.getCreatedAt(), bookmark.getId());
    }

    public static BookmarkCursor decode(String cursor) {
        try {
            String decoded = new String(Base64.getUrlDecoder().decode(cursor), StandardCharsets.UTF_8);
            String[] parts = decoded.split("\\|", 2);
            if (parts.length != 2) {
                throw new IllegalArgumentException("missing cursor fields");
            }
            return new BookmarkCursor(Instant.parse(parts[0]), UUID.fromString(parts[1]));
        } catch (Exception exception) {
            throw new BadRequestProblem("The cursor is malformed or unresolvable.");
        }
    }
}

File: backends/spring-kotlin/src/main/kotlin/dev/stackverse/backend/bookmark/BookmarkCursor.kt
package dev.stackverse.backend.bookmark

import dev.stackverse.backend.common.BadRequestProblem
import java.time.Instant
import java.util.Base64
import java.util.UUID

/**
 * Keyset position for the v2 listing: the `(createdAt, id)` of the last item on the
 * previous page, wrapped in base64url so clients treat it as opaque. Keyset pagination
 * is what makes v2 stable under concurrent inserts — new rows land before the cursor
 * position and cannot shift what the next page returns.
 */
data class BookmarkCursor(val createdAt: Instant, val id: UUID) {

    fun encode(): String =
        Base64.getUrlEncoder().withoutPadding().encodeToString("$createdAt|$id".toByteArray(Charsets.UTF_8))

    companion object {
        fun of(bookmark: Bookmark) = BookmarkCursor(bookmark.createdAt, bookmark.id)

        fun decode(cursor: String): BookmarkCursor = try {
            val decoded = String(Base64.getUrlDecoder().decode(cursor), Charsets.UTF_8)
            val (createdAt, id) = decoded.split('|', limit = 2).also { require(it.size == 2) }
            BookmarkCursor(Instant.parse(createdAt), UUID.fromString(id))
        } catch (e: Exception) {
            throw BadRequestProblem("The cursor is malformed or unresolvable.")
        }
    }
}
